x/stake/types/validator.go

```diff
@@ -386,25 +386,30 @@ func (v Validator) AddTokensFromDel(pool Pool, amount sdk.Int) (Validator, Pool,
 
 	// bondedShare/delegatedShare
 	exRate := v.DelegatorShareExRate()
-	amountDec := sdk.NewDecFromInt(amount)
+	if exRate.IsZero() {
+		panic("zero exRate should not happen")
+	}
 
 	if v.Status == sdk.Bonded {
-		pool = pool.looseTokensToBonded(amountDec)
+		pool = pool.looseTokensToBonded(amount)
 	}
 
-	if exRate.IsZero() {
-		panic("zero exRate should not happen")
-	}
-	v.Tokens = v.Tokens.Add(amountDec)
-	issuedShares := amountDec.Quo(exRate)
+	v.Tokens = v.Tokens.Add(amount)
+	issuedShares := sdk.NewDecFromInt(amount).Quo(exRate)
 	v.DelegatorShares = v.DelegatorShares.Add(issuedShares)
 
 	return v, pool, issuedShares
 }
 
 // RemoveDelShares removes delegator shares from a validator.
-func (v Validator) RemoveDelShares(pool Pool, delShares sdk.Dec) (Validator, Pool, sdk.Dec) {
-	issuedTokens := v.DelegatorShareExRate().Mul(delShares)
+// NOTE: because token fractions are left in the valiadator,
+//       the exchange rate of future shares of this validator can increase.
+func (v Validator) RemoveDelShares(pool Pool, delShares sdk.Dec) (Validator, Pool, sdk.Int) {
+	issuedTokensDec := v.DelegatorShareExRate().Mul(delShares)
+
+	// leave excess tokens in the validator
+	// however burn all the delegator shares
+	issuedTokens := issuedTokensDec.TruncateInt()
```

Comment:
note these fundamental changes

Reply:
at first glance this seems fine, we need to be careful about interactions with fee distribution but I don't think truncating and leaving in the validator vs. truncating and moving to the community pool (previous behaviour) changes the attack surface

Reply:
Yeah I don't think this is a problem, just wanted to highlight this change